### PR TITLE
8265705: aarch64: KlassDecodeMovk mode broken

### PR DESCRIPTION
**Pull request by tstuefe**
Hi,

may I please have a review for this aarch64-specific fix? Details see JBS issue and code comment.

I manually tested that this fix works. Patch includes a repro case.

Cheers, Thomas

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8265705](https://bugs.openjdk.java.net/browse/JDK-8265705): aarch64: KlassDecodeMovk mode broken

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to 16246025a3600ba76e2ea11d518c37c585b99845
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Nick Gasson](https://openjdk.java.net/census#ngasson) (@nick-arm - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3968/head:pull/3968` \
`$ git checkout pull/3968`

Update a local copy of the PR: \
`$ git checkout pull/3968` \
`$ git pull https://git.openjdk.java.net/jdk pull/3968/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3968`

View PR using the GUI difftool: \
`$ git pr show -t 3968`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3968.diff">https://git.openjdk.java.net/jdk/pull/3968.diff</a>

</details>

### CODE REVIEW

**Review by theRealAph — APPROVED** [review]

**Review by iklam — CHANGES_REQUESTED** [review]
Since we already have an assert to check that the range is less than 4G, can’t we set shift to zero for all platforms? BTW the assert should be changed to a guarantee so it works for product build as well.

In the future, if we want to both CDS and >4G CCS, we can revisit this.

**Comment by tstuefe** [comment]
Hi Ioi,

thank you for looking at this.

> Since we already have an assert to check that the range is less than 4G, can’t we set shift to zero for all platforms? BTW the assert should be changed to a guarantee so it works for product build as well.
> 
> In the future, if we want to both CDS and >4G CCS, we can revisit this.

I thought about this, but wanted to limit the patch to aarch64. We are closing in on JDK-17 code freeze and I am snowed in, I don't really want to change Klass pointer encoding now for all platforms.

Another reason is that as long as we have a shift, I am happy to have platforms which have shift>0 by default; if we set shift to 0 for all platforms on CDS, we don't have any meaningful code paths with shift >0 and it may bitrot.

If we decide to just do away with the shift altogether, ignoring the potential benefits for Lilliput, then we can do this. And if we manage to make this coding CDS agnostic, this branch will go away and we will do what the CDS=off path does today: use base==0 for pointers < 32g.

Cheers, Thomas

**Comment by iklam** [comment]
> Hi Ioi,
> 
> thank you for looking at this.
> 
> > Since we already have an assert to check that the range is less than 4G, can’t we set shift to zero for all platforms? BTW the assert should be changed to a guarantee so it works for product build as well.
> > In the future, if we want to both CDS and >4G CCS, we can revisit this.
> 
> I thought about this, but wanted to limit the patch to aarch64. We are closing in on JDK-17 code freeze and I am snowed in, I don't really want to change Klass pointer encoding now for all platforms.
> 
> Another reason is that as long as we have a shift, I am happy to have platforms which have shift>0 by default; if we set shift to 0 for all platforms on CDS, we don't have any meaningful code paths with shift >0 and it may bitrot.
> 
> If we decide to just do away with the shift altogether, ignoring the potential benefits for Lilliput, then we can do this. And if we manage to make this coding CDS agnostic, this branch will go away and we will do what the CDS=off path does today: use base==0 for pointers < 32g.
> 
> Cheers, Thomas

My reasoning is -- this is platform independent code. If this patch works for aarch64, it must also work for other platforms. Conversely, If it doesn't work for any other platform, why do we believe it would work on aarch64?

Since we test other platforms much more than aarch64, I think the best way to ensure that the patch works is to also turn it on for all platforms.

JDK 17 repo branch happens next month, but we still have a couple of months afterwards for bug fixes. I think we have enough time to handle any problems that may happen.

**Comment by tstuefe** [comment]
> > Hi Ioi,
> > thank you for looking at this.
> > > Since we already have an assert to check that the range is less than 4G, can’t we set shift to zero for all platforms? BTW the assert should be changed to a guarantee so it works for product build as well.
> > > In the future, if we want to both CDS and >4G CCS, we can revisit this.
> > 
> > 
> > I thought about this, but wanted to limit the patch to aarch64. We are closing in on JDK-17 code freeze and I am snowed in, I don't really want to change Klass pointer encoding now for all platforms.
> > Another reason is that as long as we have a shift, I am happy to have platforms which have shift>0 by default; if we set shift to 0 for all platforms on CDS, we don't have any meaningful code paths with shift >0 and it may bitrot.
> > If we decide to just do away with the shift altogether, ignoring the potential benefits for Lilliput, then we can do this. And if we manage to make this coding CDS agnostic, this branch will go away and we will do what the CDS=off path does today: use base==0 for pointers < 32g.
> > Cheers, Thomas
> 
> My reasoning is -- this is platform independent code. If this patch works for aarch64, it must also work for other platforms. Conversely, If it doesn't work for any other platform, why do we believe it would work on aarch64?
> 

Okay, I'll do it.

> Since we test other platforms much more than aarch64, I think the best way to ensure that the patch works is to also turn it on for all platforms.
> 
> JDK 17 repo branch happens next month, but we still have a couple of months afterwards for bug fixes. I think we have enough time to handle any problems that may happen.
> 
> Lilliput is not going to happen for a while, and we can handle it when the time comes. (It's not like the existing hard-coded shift of 3 is going to be much more compatible with Liliput than 0).

No, not a shift of 3. But I think a much larger shift may be interesting again, e.g. 8. This would require a couple of adjustments, but could shrink the pointer to 24 bits acc. to my thinking.

**Comment by tstuefe** [comment]
I set the shift to zero. I will hold off integrating this change to wait for test results on all our platforms.

**Review by nick-arm — APPROVED** [review]
I've tested tier1 on AArch64.

**Comment by tstuefe** [comment]
@iklam Ioi, are you fine with this?

**Review by iklam — APPROVED** [review]
LGTM.

File: src/hotspot/share/oops/compressedOops.cpp [hunk 1]
```diff
@@ -219,7 +219,16 @@ void CompressedKlassPointers::initialize(address addr, size_t len) {
     //  cannot be larger than that).
 
     base = addr;
-    shift = LogKlassAlignmentInBytes;
+
```
Comment (yminqi) [inline src/hotspot/share/oops/compressedOops.cpp:222]: Could you fix the comments at line 212, the words used seem not correct.

**Pull request closed without merging**